electroncash/lns.py

```diff
@@ -0,0 +1,607 @@
+# inherits from address.py's own ScriptOutput), so always import this file
+# carefully if also importing address.py.
+'''
+from PyQt5.QtCore import QTimer
+
+import json
+import re
+from sqlite3 import Date
+from numpy import isin
+import requests
+import threading
+import queue
+import random
+import time
+from collections import defaultdict, namedtuple
+from typing import List, Tuple, Dict, Union
+from . import util
+from .address import Address, Base58, OpCodes, Script, ScriptError, UnknownAddress
+from .address import ScriptOutput as ScriptOutputBase
+from .transaction import BCDataStream, Transaction, get_address_from_output_script
+
+# 'lns:' URI scheme. Not used yet. Used by Crescent Cash and Electron Cash and
+# other wallets in the future.
+URI_SCHEME = 'lns'
+
+class ArgumentError(ValueError):
+    '''Raised by various LNS functions if the supplied args are bad or
+    out of spec.'''
+
+#### Lookup & Verification
+
+class Info(namedtuple("Info", "name, address, registrationDate, expiryDate")):
+    @classmethod
+    def from_dict(cls, dict):
+        dict['address'] = Address.from_cashaddr_string(dict['address'])
+        tup = Info(**dict)
+        return tup
+
+    def to_dict(self):
+        d = self._asdict()
+        d['address'] = self.address.to_cashaddr()
+        return d
+
+servers = [
+    "https://graph.bch.domains/subgraphs/name/graphprotocol/ens",
+]
+
+debug = False  # network debug setting. Set to True when developing to see more verbose information about network operations.
+timeout = 25.0  # default timeout used in various network functions, in seconds.
+
+abi = [
+    {
+      "inputs": [
+        {
+          "internalType": "contract ENS",
+          "name": "_ens",
+          "type": "address"
+        }
+      ],
+      "stateMutability": "nonpayable",
+      "type": "constructor"
+    },
+    {
+      "inputs": [
+        {
+          "internalType": "string[]",
+          "name": "names",
+          "type": "string[]"
+        },
+        {
+          "internalType": "uint256",
+          "name": "coinType",
+          "type": "uint256"
+        }
+      ],
+      "name": "getAddrs",
+      "outputs": [
+        {
+          "internalType": "bytes[]",
+          "name": "r",
+          "type": "bytes[]"
+        }
+      ],
+      "stateMutability": "view",
+      "type": "function"
+    },
+    {
+      "inputs": [
+        {
+          "internalType": "address[]",
+          "name": "addresses",
+          "type": "address[]"
+        }
+      ],
+      "name": "getNames",
+      "outputs": [
+        {
+          "internalType": "string[]",
+          "name": "r",
+          "type": "string[]"
+        }
+      ],
+      "stateMutability": "view",
+      "type": "function"
+    }
+  ]
+from web3 import Web3
+w3 = Web3(Web3.HTTPProvider("https://smartbch.fountainhead.cash/mainnet"))
+contract = w3.eth.contract(address="0x0efB8EE0F6d6ba04F26101683F062d7Ca6F58A40", abi=abi)
```

Comment:
comment describing contract and link to code repo would make it easier to understand

Reply:
commented